app/controllers/sites/new.js

```diff
@@ -58,52 +58,83 @@ export default class CreateSitesNewController extends Controller {
     };
   }
 
+  mapValidationDetailsToErrors(validationDetails) {
+    return validationDetails.reduce((accumulator, detail) => {
+      accumulator[detail.context.key] = detail.message;
+      return accumulator;
+    }, {});
+  }
+
+  async saveFormData() {
+    const { address, primaryContact, secondaryContact, site, adminUnit } =
+      this.model;
+    address.fullAddress = combineFullAddress(address);
+
+    await primaryContact.save();
+    await secondaryContact.save();
+    await address.save();
+
+    site.contacts = [primaryContact, secondaryContact];
+    site.address = address;
+    await site.save();
+
+    const nonPrimarySites = await adminUnit.sites;
+
+    if (this.isPrimarySite) {
+      const previousPrimarySite = await adminUnit.primarySite;
+
+      if (previousPrimarySite) {
+        nonPrimarySites.push(previousPrimarySite);
+      }
+
+      adminUnit.primarySite = site;
+    } else {
+      nonPrimarySites.push(site);
+    }
+
+    await adminUnit.save();
+  }
+
   saveTask = task(async (event) => {
     event.preventDefault();
-    this.validationErrors = {};
-    this.validationWarnings = {};
-    const validationResult = await this.validateData();
+    const validationResult = await this.validateFormData();
+
     if (
       validationResult.errors &&
       Object.keys(validationResult.errors).length > 0
     ) {
-      // Handle errors
-      this.validationErrors = validationResult.errors.details.reduce(
-        (errors, detail) => {
-          errors[detail.context.key] = detail.message;
-          return errors;
-        },
-        {},
+      this.validationErrors = this.mapValidationDetailsToErrors(
+        validationResult.errors.details,
       );
+      console.log(validationResult.errors.details);
       return;
     }
+
     if (
       validationResult.warnings &&
       validationResult.warnings.details.length > 0
     ) {
-      console.log('er zijn warnings');
+      console.log('There are warnings');
       this.showWarningModal = true;
-      this.validationWarnings = validationResult.warnings.details.reduce(
-        (warnings, detail) => {
-          warnings[detail.context.key] = detail.message;
-          return warnings;
-        },
-        {},
+      this.validationWarnings = this.mapValidationDetailsToErrors(
+        validationResult.warnings.details,
       );
       return;
     }
+
     // No errors and no warnings
-    await this.save();
+    await this.saveFormData();
     this.reset();
     this.router.transitionTo('sites.index');
   });
 
-  handleWarningModalOK = task(async (event) => {
+  @action
+  async handleWarningModalOK(event) {
```

Comment:
This action handler does not allow for a correct loading animation.
It should launch a save task and this task must do the redirecting. Not this handler.
Now the application will look like 't crashed/frozen to the user when the saving process works.
Here too something like `this.saveTask.perform();` will need to be added. Close the form using this.showWarningModal = false so the user can
1. See that his input is received
2. See the loading animation(s) on the page.